transport/http2_client.go

```diff
@@ -114,11 +115,32 @@ type http2Client struct {
 	prevGoAwayID uint32
 }
 
-func dial(ctx context.Context, fn func(context.Context, string) (net.Conn, error), addr string) (net.Conn, error) {
-	if fn != nil {
-		return fn(ctx, addr)
+func dial(ctx context.Context, fn func(context.Context, string) (net.Conn, error), p proxy.Proxyer, addr string) (net.Conn, error) {
+	if fn == nil {
+		fn = func(ctx context.Context, addr string) (net.Conn, error) {
+			return dialContext(ctx, "tcp", addr)
+		}
+	}
+
+	if p == nil {
+		p = proxy.NewEnvironmentVariableHTTPConnectProxy()
+	}
+
+	var skipHandshake bool
+	newAddr, h, err := p.MapAddress(ctx, addr)
+	if err != nil {
+		if err != proxy.ErrIneffective {
+			return nil, err
+		}
+		skipHandshake = true
+		newAddr = addr
+	}
+
+	conn, err := fn(ctx, newAddr)
```

Comment:
What if the proxyer was responsible for connecting?  In that case, the proxyer would be mostly a wrapper, and the logic around here could be completely offloaded to it.  I.e.:

```
package proxy

type Dialer interface {
  Dial(ctx context.Context, addr string, dialFunc func(context.Context, string) (net.Conn, error)) (net.Conn, error)
}
```

A proxy's Dial method would dial, determine the headers and new address if necessary, and perform whatever handshaking is required.  If the proxy was disabled/ineffective, it would be responsible for doing simply: "return dialFunc(ctx, addr)".  This avoids the need to pass around the headers / new address / disabled error, and should allow more flexibility in what Proxies could implement.